File: .claude/skills/cernji-logging/assets/src/cernji-logging-ts/src/timing.ts
/**
 * Performance timing utilities for logging
 */

import type { Logger } from 'pino';

/**
 * Class method decorator to log execution time
 *
 * @param level - Log level to use (default: 'info')
 *
 * @example
 * ```typescript
 * class MyService {
 *   @Timed()
 *   async expensiveOperation() {
 *     // ... code ...
 *   }
 * }
 * ```
 */
export function Timed(level: 'trace' | 'debug' | 'info' | 'warn' | 'error' = 'info') {
  return function (
    target: any,
    propertyKey: string,
    descriptor: PropertyDescriptor
  ) {
    const originalMethod = descriptor.value;

    descriptor.value = async function (this: any, ...args: any[]) {
      const startTime = performance.now();
      const logger = this.logger as Logger | undefined;

      try {
        const result = await originalMethod.apply(this, args);
        return result;
      } finally {
        const durationMs = performance.now() - startTime;

        if (logger && typeof logger[level] === 'function') {
          logger[level](
            {
              function: propertyKey,
              duration_ms: durationMs,
              duration_ns: durationMs * 1_000_000,
            },
            `Function ${propertyKey} completed`
          );
        }
      }
    };

    return descriptor;
  };
}

/**
 * Manual timing context for code blocks
 *
 * @example
 * ```typescript
 * const timer = new TimingContext(logger, 'database_query');
 * // ... code to time ...
 * timer.end(); // Logs duration
 * ```
 */
export class TimingContext {
  private startTime: number;
  private durationMs: number = 0;

  constructor(
    private logger: Logger,
    private operation: string,
    private level: 'trace' | 'debug' | 'info' | 'warn' | 'error' = 'info'
  ) {
    this.startTime = performance.now();
  }

  /**
   * End timing and log duration
   */
  end(error?: Error): void {
    this.durationMs = performance.now() - this.startTime;

    if (typeof this.logger[this.level] === 'function') {
      this.logger[this.level](
        {
          operation: this.operation,
          duration_ms: this.durationMs,
          duration_ns: this.durationMs * 1_000_000,
          error: error !== undefined,
        },
        `Operation ${this.operation} completed`
      );
    }
  }

  /**
   * Get the current duration without logging
   */
  getDuration(): number {
    return performance.now() - this.startTime;
  }
}

/**
 * Timing middleware for HTTP frameworks (Express, Hono, etc.)
 *
 * @example
 * ```typescript
 * // Express
 * app.use(timingMiddleware(logger));
 *
 * // Hono
 * app.use(timingMiddleware(logger));
 * ```
 */
export function timingMiddleware(logger: Logger) {
  return async (req: any, res: any, next: any) => {
    const startTime = performance.now();
    const path = req.path || req.url;
    const method = req.method;

    // Add timing to response finish
    const originalEnd = res.end;
    res.end = function (this: any, ...args: any[]) {
      const durationMs = performance.now() - startTime;

      logger.info(
        {
          http: {
            request: {
              method,
              path,
            },
            response: {
              status_code: res.statusCode,
            },
          },
          duration_ms: durationMs,
          duration_ns: durationMs * 1_000_000,
        },
        `${method} ${path} ${res.statusCode}`
      );

      return originalEnd.apply(this, args);
    };

    if (next) {
      return next();
    }
  };
}

File: libs/cernji-logging-py/src/cernji_logging/timing.py
"""Performance timing utilities for logging."""

import functools
import time
from typing import Any, Callable, TypeVar

from structlog.stdlib import BoundLogger

F = TypeVar("F", bound=Callable[..., Any])


def timed(logger: BoundLogger, level: str = "info"):
    """Decorator to log function execution time.

    Args:
        logger: The logger to use for timing logs.
        level: Log level to use (debug, info, warn, error).

    Example:
        ```python
        @timed(logger)
        def expensive_function():
            # ... slow code ...
            pass

        @timed(logger, level="debug")
        async def async_function():
            # ... async code ...
            pass
        ```
    """

    def decorator(func: F) -> F:
        log_method = getattr(logger, level, logger.info)

        @functools.wraps(func)
        def sync_wrapper(*args, **kwargs):
            start_time = time.perf_counter_ns()
            try:
                result = func(*args, **kwargs)
                return result
            finally:
                duration_ns = time.perf_counter_ns() - start_time
                log_method(
                    f"Function {func.__name__} completed",
                    function=func.__name__,
                    module=func.__module__,
                    duration_ns=duration_ns,
                    duration_ms=duration_ns / 1_000_000,
                )

        @functools.wraps(func)
        async def async_wrapper(*args, **kwargs):
            start_time = time.perf_counter_ns()
            try:
                result = await func(*args, **kwargs)
                return result
            finally:
                duration_ns = time.perf_counter_ns() - start_time
                log_method(
                    f"Function {func.__name__} completed",
                    function=func.__name__,
                    module=func.__module__,
                    duration_ns=duration_ns,
                    duration_ms=duration_ns / 1_000_000,
                )

        # Return appropriate wrapper based on whether function is async
        if asyncio_iscoroutinefunction(func):
            return async_wrapper  # type: ignore
        else:
            return sync_wrapper  # type: ignore

    return decorator


def asyncio_iscoroutinefunction(func):
    """Check if a function is an async coroutine function.

    Uses inspect.iscoroutinefunction but handles edge cases.
    """
    import inspect

    return inspect.iscoroutinefunction(func)


class TimingContext:
    """Context manager for timing code blocks.

    Example:
        ```python
        with TimingContext(logger, "database_query") as timer:
            # ... code to time ...
            pass
        # Automatically logs duration
        ```
    """

    def __init__(self, logger: BoundLogger, operation: str, level: str = "info"):
        """Initialize timing context.

        Args:
            logger: The logger to use.
            operation: Name of the operation being timed.
            level: Log level to use.
        """
        self.logger = logger
        self.operation = operation
        self.level = level
        self.start_time: int = 0
        self.duration_ns: int = 0

    def __enter__(self):
        """Start timing."""
        self.start_time = time.perf_counter_ns()
        return self

    def __exit__(self, exc_type, exc_val, exc_tb):
        """Stop timing and log duration."""
        self.duration_ns = time.perf_counter_ns() - self.start_time
        log_method = getattr(self.logger, self.level, self.logger.info)

        log_method(
            f"Operation {self.operation} completed",
            operation=self.operation,
            duration_ns=self.duration_ns,
            duration_ms=self.duration_ns / 1_000_000,
            error=exc_type is not None,
        )
